mixnet/packet.py

```diff
@@ -0,0 +1,203 @@
+
+    header: FragmentHeader
+    body: bytes
+
+    @staticmethod
+    def max_payload_size() -> int:
+        return Payload.max_plain_payload_size() - FragmentHeader.size()
+
+    def bytes(self) -> bytes:
+        return self.header.bytes() + self.body
+
+    @classmethod
+    def from_bytes(cls, data: bytes) -> Self:
+        header = FragmentHeader.from_bytes(data[: FragmentHeader.size()])
+        body = data[FragmentHeader.size() :]
+        return cls(header, body)
+
+
+# Unlikely, Nym uses i32 for FragmentSetId, which may cause more collisions.
+# We will use UUID until figuring out why Nym uses i32.
+FragmentSetId: TypeAlias = bytes  # 128bit UUID v4
+FragmentId: TypeAlias = int  # unsigned 8bit int in big endian
+
+
+@dataclass
+class FragmentHeader:
+    """
+    Contain all information for reconstructing a message that was fragmented into the same FragmentSet.
+    """
+
+    set_id: FragmentSetId
+    total_fragments: FragmentId
+    fragment_id: FragmentId
+
+    @staticmethod
+    def size() -> int:
+        return 16 + 1 + 1
+
+    @staticmethod
+    def max_total_fragments() -> int:
+        return 256  # because total_fragment is u8
+
+    def bytes(self) -> bytes:
+        return (
+            self.set_id
+            + self.total_fragments.to_bytes(1)
+            + self.fragment_id.to_bytes(1)
+        )
+
+    @classmethod
+    def from_bytes(cls, data: bytes) -> Self:
+        assert len(data) == cls.size()
+        return cls(data[:16], int.from_bytes(data[16:17]), int.from_bytes(data[17:18]))
+
+
+@dataclass
+class MessageReconstructor:
+    fragmentSets: Dict[FragmentSetId, FragmentSetReconstructor]
+
+    @classmethod
+    def new(cls) -> Self:
+        return cls({})
+
+    def add(self, fragment: Fragment) -> bytes | None:
+        if fragment.header.set_id not in self.fragmentSets:
+            self.fragmentSets[fragment.header.set_id] = FragmentSetReconstructor.new(
+                fragment.header.total_fragments
+            )
+
+        msg = self.fragmentSets[fragment.header.set_id].add(fragment)
+        if msg is not None:
+            del self.fragmentSets[fragment.header.set_id]
+        return msg
+
+
+@dataclass
+class FragmentSetReconstructor:
+    total_fragments: FragmentId
+    fragments: Dict[FragmentId, Fragment]
+
+    @classmethod
+    def new(cls, total_fragments: FragmentId) -> Self:
+        return cls(total_fragments, {})
+
+    def add(self, fragment: Fragment) -> bytes | None:
+        self.fragments[fragment.header.fragment_id] = fragment
+        if len(self.fragments) == self.total_fragments:
+            message = b""
+            for i in range(self.total_fragments):
+                message += self.fragments[FragmentId(i)].body
+            return message
+        else:
+            return None
+
+
+def chunks(data: bytes, size: int) -> List[bytes]:
+    return [data[i : i + size] for i in range(0, len(data), size)]
```

Comment:
nitpick: This is the same as latest [`batched`](https://docs.python.org/3/library/itertools.html#itertools.batched).

Reply:
Wow. I didn't know itertools has it. How do you remember this latest change? :)
d995935 Here, I deliberately flatten an iterator, returned from `batched` to a list, in order to clearly show that the number of chunks shouldn't exceed the limitation.